File: week5/InheritanceTypes.cpp
#include<iostream>
using namespace std;

class Class1
{
    public:
        void class1()
        {
            cout<<"Iam a Class1\n";
        }
};

class Class2
{
    public:
        void class2()
        {
            cout<<"Iam a Class2\n";
        }
};

class Class3: public Class1,public Class2
{
    public:
        void class3()
        {
            cout<<"Iam a Class3\n";
        }
};

class Class4: public Class1
{
    public:
        void class4()
        {
            cout<<"Iam a Class4\n";
        }
};

class Class5: public Class3,public Class4
{
    public:
        void class5()
        {
            cout<<"Iam a Class5\n";
        }
};

class Class6: public Class4
{
    public:
        void class6()
        {
            cout<<"Iam a Class6\n";
        }
};

class Class7:public Class1
{
    public:
        void class7()
        {
            cout<<"Iam a Class7\n";
        }
};
class Class8:public Class1
{
    public:
        void class8()
        {
            cout<<"Iam a Class8\n";
        }
};

int main()
{
    cout<<"Simple Inheritance\n";
    Class4 obj1;
    obj1.class1();
    obj1.class4();
    cout<<"Multiple Inheritance\n";
    Class3 obj2;
    obj2.class1();
    obj2.class2();
    obj2.class3();
    cout<<"Multi-Level Inheritance\n";
    Class6 obj3;
    obj3.class1();
    obj3.class4();
    obj3.class6();
    cout<<"Hierarichal Inheritance\n";
    Class7 obj4;
    Class8 obj5;
    obj4.class1();
    obj4.class7();
    obj5.class1();
    obj5.class8();
    cout<<"Hybrid Inheritance\n";
    Class5 obj6;
    obj6.class1();
    //class1() method is ambiguous here.
    obj6.class2();
    obj6.class3();
    obj6.class4();
    obj6.class5();
    return 0;
}

File: week5/InheritanceTypes.java
class Class1
{
    public
        void class1()
        {
            System.out.print("Iam a class1"+"\n");
        }
}

class Class2 extends Class1
{
    public
        void class2()
        {
            System.out.print("Iam a class2"+"\n");
        }
}

class Class3 extends Class1
{
    public
        void class3()
        {
            System.out.print("Iam a class3"+"\n");
        }
}

class Class4 extends Class2
{
    public
        void class4()
        {
            System.out.print("Iam a class4"+"\n");
        }
}

class Class5 extends Class1,Class2
{
    
}
/*Java doesnt support Multiple Inheritance and Hybrid Inheritance.*/

public class InheritanceTypes
{
    public static void main(String[] args)
    {
        System.out.print("Simple Inheritance"+"\n");
        Class2 class1=new Class2();
        class1.class1();
        class1.class2();
        System.out.print("Hierarichal Inheritance"+"\n");
        Class3 class2=new Class3();
        class1.class1();
        class1.class2();
        class2.class1();
        class2.class3();
        System.out.print("Multi-Level Inheritance"+"\n");
        Class4 class3=new Class4();
        class3.class1();
        class3.class2();
        class3.class4();
        /*Java does not support multiple inheritances to avoid the diamond problem 
        which causes complexity and ambiguity. Well we can acheive multiple inheritance by the concept of interface.*/
    }
}
